File: src/otherAlgorithms/gcdandlcm.ts
/**
 * Calculates the greatest common divisor (GCD) of two numbers.
 *
 * @param {number} n1 - The first number.
 * @param {number} n2 - The second number.
 * @returns {number} - The GCD of the two numbers.
 */
function gcd(n1: number, n2: number): number {
  if (n2 === 0) {
    return n1
  } else {
    return gcd(n2, n1 % n2)
  }
}

/**
 * Calculates the least common multiple (LCM) of two numbers.
 *
 * @param {number} n1 - The first number.
 * @param {number} n2 - The second number.
 * @returns {number} - The LCM of the two numbers.
 */
function lcm(n1: number, n2: number): number {
  return (n1 * n2) / gcd(n1, n2)
}

/**
 * Calculates the GCD and LCM of two numbers.
 *
 * @param {number} n1 - The first number.
 * @param {number} n2 - The second number.
 * @returns {Object} - The GCD and LCM as properties of an object.
 */
function calculateGCDAndLCM(
  n1: number,
  n2: number
): { gcd: number; lcm: number } {
  if (n1 === null || n2 === null || isNaN(n1) || isNaN(n2)) {
    throw new Error("Invalid input: number must be a integer.")
  }

  const gcdResult = gcd(Math.abs(n1), Math.abs(n2))
  const lcmResult = lcm(Math.abs(n1), Math.abs(n2))

  console.log(`The greatest common divisor of ${n1} and ${n2} is ${gcdResult}`)
  console.log(`The least common multiple of ${n1} and ${n2} is ${lcmResult}`)

  return { gcd: gcdResult, lcm: lcmResult }
}

export { calculateGCDAndLCM }

// Example usage
// console.log(calculateGCDAndLCM(12, 21))

File: src/otherAlgorithms/gcdandlcm.c
#include <stdio.h>

long gcd(long, long);
long lcm(long, long);

int main()
{
  long n1, n2, hcf, lcmValue;

  printf("Enter two integers: ");
  scanf("%ld%ld", &n1, &n2);

  // Check if either number is negative. If so, make it positive.
  if (n1 < 0)
  {
    n1 = -n1;
  }
  if (n2 < 0)
  {
    n2 = -n2;
  }

  // Find the HCF and LCM using the gcd() and lcm() functions.
  hcf = gcd(n1, n2);
  lcmValue = lcm(n1, n2);

  // Print the results.
  printf("Greatest common divisor of %ld and %ld = %ld\n", n1, n2, hcf);
  printf("Least common multiple of %ld and %ld = %ld\n", n1, n2, lcmValue);

  return 0;
}

// Recursive function to find the greatest common divisor (GCD) of two numbers.
long gcd(long n1, long n2)
{
  if (n2 == 0)
  {
    return n1;
  }
  else
  {
    return gcd(n2, n1 % n2);
  }
}

// Function to find the least common multiple (LCM) of two numbers.
long lcm(long n1, long n2)
{
  return n1 * n2 / gcd(n1, n2);
}
